### DISCUSSION TITLE
Windows runtime error and Izhikevich neuron model

### DISCUSSION THREAD

**Discussion by ahmetozleraktas** [lava-nc/lava#257] [Q&A] [answered]
Hello, I hope you're all doing well. I am trying to solve this problem which I mentioned in the title. There is no ready-to-use model for Izhikevich neuron model. I saw the LIF but i couldn't manage to modify it for changing to Izhikevich model. Are there any documentation for it? Can you help me? Thanks for all answers! Have a nice day.

**Comment by weidel-p** [comment]
Hi,

you are right, there is no ready to use Izhikevich neuron model available in the process library yet. Have you seen the jupyter notebook tutorials? Especially the tutorial about creating custom [ProcessModels](https://github.com/lava-nc/lava/blob/main/tutorials/in_depth/tutorial03_process_models.ipynb) from scratch should be helpful for you.

Please don't hesitate to ask if you need further assistance. If you succeed in implementing the Izhikevic neuron model, consider making it available to other users by creating a pull request.

**Reply by ahmetozleraktas** [reply]
I checked the MNIST classifier, but didn't notice that tutorial you mentioned. It seems a bit difficult to begin but maybe I can figure it out. You can be sure, I will add the model if I solve it. Thanks for your answer by the way. Best regards!

**Comment by felwang-sandia** [comment]
Did this ever get resolved/added?

By the way, a simple izhikevich model I've used previously (taken from [1]) would have the population-wide (constant) parameters: a, b, c, d, and the individual (mutable) states: v, u, I.
The forward equations would be (time-step split into two):
v = v + dt/2 \* ((0.04\*v + 5)\*v + 140 - u + I
v = v + dt/2 \* ((0.04\*v + 5)\*v + 140 - u + I
u = u + dt\*a \* (b*v - u)
On spike:
v = c
u = u + d

[1] E. Izhikevich, Which Model to Use for Cortical Spiking Neurons? IEEE Trans. on Neural Networks, v. 15, no. 5, 2004

**Reply by awintel** [reply]
We have not created an Izhikevich neuron model ourselves so far at NCL. 
Are you asking about a Python or NeuroCore model? Certainly something like an Izhikevich model can be coded up both as a PyProcModel and as an NcProcModel.

**Reply by felwang-sandia** [reply]
Thanks for the quick reply. I don't personally have an immediate use for it, but I think both would eventually be nice to have for reservoir computing models (e.g. something like [1]). It would also be nice to reproduce the E-I network in [2] as a toy model, as it can act as the basis for more complex bio-inspired models looking at multiple connected E-I populations.

[1] H. Paugam-Moisy, et. al. Delay learning and polychronization for reservoir computing, Neurocomputing, v. 71, 2008
[2] E. Izhikevich, Polychronization: computation with spikes, Neural Computation, v. 18, no. 2, 2006

**Reply by ahmetozleraktas** [reply]
We were able to test the Izhikevich neurons with different neuron types. But not yet on Loihi. It's partially running and we're getting printouts, but the simulations are taking too long. When I take the time step to 0.1 and the simulation time to 200 milliseconds, I get results, but it is very difficult to reduce time step because simulations take too long. Do you have an idea about it?

**Reply by awintel** [reply]
Would you be able to share the model? We might be able to take a look at it and see where the slowdown comes from.

**Reply by felwang-sandia** [reply]
For the model equations that I wrote out above at least, I think the default timestep is usually 1ms. Do you also have a "current generator" model connecting to the Izhikevich neuron model that you're using for testing?

**Reply by ahmetozleraktas** [reply]
> For the model equations that I wrote out above at least, I think the default timestep is usually 1ms. Do you also have a "current generator" model connecting to the Izhikevich neuron model that you're using for testing?

```

a_in_data = self.a_in.recv()
        self.v = self.v + (0.04*self.v**2 + 5*self.v + 140 - self.u + self.I)*self.step_size
        self.u = self.u + (self.a)*((self.b*self.v - self.u))*self.step_size
        s_out = self.v >= self.vth # spike condition
        self.v[s_out] = self.c[s_out] # Reset voltage to c value
        self.u[s_out] = self.u[s_out] + self.d[s_out] # Reset u to d value
        self.s_out.send(s_out)

```
I am taking step size and time step as input variables. Process input and outputs are empty for right now.

**Reply by awintel** [reply]
To clarify: Could you share an entire test case based on which you determined that it's too slow?
You could just create an issue explaining what we need to know about the test case together with the test case code.
Thanks

**Reply by felwang-sandia** [reply]
I cobbled this together just now by modifying some of the tutorial cells (~~i.e. there are parts of it that are wrong and still need modification, such as v_init, u_init being set to zero and ignoring user-supplied values~~ updated to hopefully be correct). It has a bias/current of 10 which lets it spike about 4 times over 200 timesteps at 1ms/timestep. Edit: also added some simple timing code.

```
import numpy as np
from lava.magma.core.process.process import AbstractProcess
from lava.magma.core.process.variable import Var
from lava.magma.core.process.ports.ports import InPort, OutPort

class Izhi(AbstractProcess):
    """Leaky-Integrate-and-Fire (LIF) neural Process.

    """
    def __init__(self, **kwargs):
        super().__init__()
        shape = kwargs.get("shape", (1,))
        a = kwargs.pop("a", 0.02)
        b = kwargs.pop("b", 0.2)
        c = kwargs.pop("c", -65.0)
        d = kwargs.pop("d", 8.0)
        bias_mant = kwargs.pop("bias_mant", 0)
        bias_exp = kwargs.pop("bias_exp", 0)
        u = kwargs.pop("u", -14.0)
        v = kwargs.pop("v", -70.0)
        vth = kwargs.pop("vth", 30.0)
        dt = kwargs.pop("dt", 1.0)

        self.shape = shape
        self.a_in = InPort(shape=shape)
        self.s_out = OutPort(shape=shape)
        self.u = Var(shape=shape, init=u)
        self.v = Var(shape=shape, init=v)
        self.I = Var(shape=shape, init=0)
        self.a = Var(shape=(1,), init=a)
        self.b = Var(shape=(1,), init=b)
        self.c = Var(shape=(1,), init=c)
        self.d = Var(shape=(1,), init=d)
        self.bias_mant = Var(shape=shape, init=bias_mant)
        self.bias_exp = Var(shape=shape, init=bias_exp)
        self.vth = Var(shape=(1,), init=vth)
        self.dt = Var(shape=(1,), init=dt)
        
from lava.magma.core.sync.protocols.loihi_protocol import LoihiProtocol
from lava.magma.core.model.py.ports import PyInPort, PyOutPort
from lava.magma.core.model.py.type import LavaPyType
from lava.magma.core.resources import CPU
from lava.magma.core.decorator import implements, requires, tag
from lava.magma.core.model.py.model import PyLoihiProcessModel

@implements(proc=Izhi, protocol=LoihiProtocol)
@requires(CPU)
@tag('floating_pt')
class PyIzhiModel1(PyLoihiProcessModel):
    a_in: PyInPort = LavaPyType(PyInPort.VEC_DENSE, float)
    s_out: PyOutPort = LavaPyType(PyOutPort.VEC_DENSE, bool, precision=1)
    u: np.ndarray = LavaPyType(np.ndarray, float)
    v: np.ndarray = LavaPyType(np.ndarray, float)
    I: np.ndarray = LavaPyType(np.ndarray, float)
    bias_mant: np.ndarray = LavaPyType(np.ndarray, float)
    bias_exp: np.ndarray = LavaPyType(np.ndarray, float)
    a: float = LavaPyType(float, float)
    b: float = LavaPyType(float, float)
    c: float = LavaPyType(float, float)
    d: float = LavaPyType(float, float)
    vth: float = LavaPyType(float, float)
    dt: float = LavaPyType(float, float)

    def run_spk(self):
        a_in_data = self.a_in.recv()
        self.I[:] += a_in_data
        bias = self.bias_mant * (2 ** self.bias_exp)
        self.v[:] = self.v + (0.04*self.v**2 + 5*self.v + 140 - self.u + self.I + bias)*self.dt/2
        self.v[:] = self.v + (0.04*self.v**2 + 5*self.v + 140 - self.u + self.I + bias)*self.dt/2
        self.u[:] = self.u + (self.a)*((self.b*self.v - self.u))*self.dt
        self.I[:] = 0 # Reset inputs to zero (TODO: probably want to use a synaptic time constant here)
        s_out = self.v >= self.vth # spike condition
        self.v[s_out] = self.c # Reset voltage to c value
        self.u[s_out] = self.u[s_out] + self.d # Reset u to d value
        self.s_out.send(s_out)

from lava.magma.core.run_configs import Loihi1SimCfg
from lava.magma.core.run_conditions import RunSteps
from lava.proc.monitor.process import Monitor
import time

izhi = Izhi(shape=(1,), dt=1.0, u=0, v=-65, bias_mant=10, vth=30)

num_steps = 200
monitor_izhi = Monitor()
monitor_izhi.probe(izhi.v, num_steps)

run_cfg = Loihi1SimCfg()
tic = time.time()
izhi.run(condition=RunSteps(num_steps=num_steps), run_cfg=run_cfg)
toc = time.time()
print(toc - tic)

data_izhi = monitor_izhi.get_data()

import matplotlib
%matplotlib inline
from matplotlib import pyplot as plt

# Create a subplot for each monitor
fig = plt.figure(figsize=(10, 5))
ax0 = fig.add_subplot(111)

# Plot the recorded data
monitor_izhi.plot(ax0, izhi.v)
```

**Reply by awintel** [reply]
Thanks. For completeness, could you also provide what you measure in terms of runtime on your end and what you would expect in terms of runtime?
Your numpy code should not be the issue. It's likely something in the channel communication.

If you don't mind, please open an Issue for this. This way we can communicate on this particular speed problem and close it when the problem is resolved.

**Reply by felwang-sandia** [reply]
I updated the code above to add some timing. I'm not sure how long it's taking on Amhet's side, but 200 time steps on my end is about 0.4s with the monitor, and about 0.1s without, which seems roughly in line with what I experienced when running the LIF neurons in the tutorial. I don't really have a baseline of what to expect for the LIF neurons on CPU, so I don't know if I'm having the same issue that Amhet is having.

**Reply by ahmetozleraktas** [reply]
> I updated the code above to add some timing. I'm not sure how long it's taking on Amhet's side, but 200 time steps on my end is about 0.4s with the monitor, and about 0.1s without, which seems roughly in line with what I experienced when running the LIF neurons in the tutorial. I don't really have a baseline of what to expect for the LIF neurons on CPU, so I don't know if I'm having the same issue that Amhet is having.

Thanks for all of your effort. I tried to run your code but it gave the following error.

Edit: Sorry that's my fault, I can not run this code on Jupyter Notebook, it really takes too much time and I can not review any output (beacuse runtime goes to infinity). So I need to run this code on a .py file.

**Reply by ahmetozleraktas** [reply]
> Thanks. For completeness, could you also provide what you measure in terms of runtime on your end and what you would expect in terms of runtime? Your numpy code should not be the issue. It's likely something in the channel communication.
> 
> If you don't mind, please open an Issue for this. This way we can communicate on this particular speed problem and close it when the problem is resolved.

You are very right, at first I thought it was about me, so I thought I could handle it from here. If I can run @felwang-sandia 's code, I will create an issue with the outputs.

**Reply by awintel** [reply]
So @felwang-sandia  can run it in jupyter (or .py?) but you, @ahmetozleraktas can't?
Maybe an operating system issue?
If you figure it out, please include that info in the issue.

**Reply by felwang-sandia** [reply]
Yes, I'm running the code on a jupyter notebook. My development environment is on an ubuntu 22.04 machine, I'm using anaconda for the python virtual environment with python 3.8, and I'm using lava 0.5 (installed with "conda install lava -c conda-forge").

**Reply by ahmetozleraktas** [reply]
@awintel I opened an issue: #401.

**Reply by awintel** [reply]
Thanks!

**Comment by awintel** [comment]
Ok. Thanks for the suggestions.
FYI: We have just released an E/I network tutorial, although not with Izhikevich neurons but normal rate and LIF neurons:
https://github.com/lava-nc/lava/blob/main/tutorials/end_to_end/tutorial02_excitatory_inhibitory_network.ipynb

**Reply by felwang-sandia** [reply]
Thanks, I've since taken a look at the E/I network tutorial. It's very nice, and I especially like the additional analysis function looking at criticality/correlation length. I know sparse connectivity and varied synaptic delays (tag_2, I believe?) are not quite implemented yet, but a network with those additional properties would be a great addition to the E/I network tutorial if/when they come about.

**Reply by awintel** [reply]
The compiler should support synaptic delays by now. But we do not have a tutorial for it yet.

**Reply by felwang-sandia** [reply]
That's great!

**Comment by ahmetozleraktas** [comment, marked as answer]
**UPDATE:** I tried the codes separately on Ubuntu and Windows. There weren't any of the problems I mentioned in Ubuntu, but Windows runtime and Jupyter Notebook problems continue. I opened an Issue about it (#401). I guess it's an operating system problem for now.
